{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Pranav-Datar/pacbio_denovo2/blob/main/NGS_colab_part3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Introduction to Variant Filtering\n",
        "\n",
        "High-throughput sequencing technologies generate millions of genetic variants across genomes, but not all detected variants are biologically meaningful or technically reliable. Raw variant calls often contain sequencing errors, low-confidence genotypes, and missing data that can severely bias downstream analyses. From removing errors to retain only high-confidence SNPs, each step here helps you build a more accurate dataset for further applications such as population genetics.\n",
        "Below are the details of the filters that we are going to use during this analysis to get a filtered VCF file.\n",
        "\n",
        "1: *passOnly*-  Retained only variant sites that passed all internal quality filter performed by the variant caller (Strelka in this case). During variant calling, Strelka assigns a PASS or FAIL tag to each variant based on multiple internal metrics such as base quality, mapping quality, strand bias, and read position bias.\n",
        "\n",
        "2: *biallelicOnly*- Kept only biallelic variants (i.e., sites with exactly two alleles) for downstream compatibility and clarity. Because most population genetic software assumes biallelic markers, this filter ensures mathematical simplicity, software compatibility, and interpretability.\n",
        "\n",
        "3: *rmvIndels*- Removed insertions and deletions (INDELs), keeping only SNPs (single nucleotide polymorphisms). Indels have higher sequencing and alignment error rates than SNPs. Removing indels ensures that the dataset is composed of high-confidence point mutations.\n",
        "\n",
        "4: *minMAF0Pt05*- Retained only variants with minor allele frequency (MAF) ≥ 0.05, meaning the alternate allele must be present in at least 5% of individuals. As very rare variants (MAF < 5%) are more likely to be sequencing errors, they are often poorly powered in statistical tests.\n",
        "\n",
        "5: *chr_E2*- Restricted to a specific chromosome/region of interest (e.g., chr_E2). This enables targeted analysis of specific genomic regions.\n",
        "\n",
        "6: *minDP3*- Required a minimum depth (DP) of 3 per genotype to avoid false positives due to low coverage. Because low-depth genotypes are highly error-prone and vulnerable to random sequencing noise.\n",
        "\n",
        "7: *minQ30*- Ensured that each site has a minimum site quality score of 30, reflecting high confidence in the variant.  This improves accuracy of allele frequency estimates and reliability of genome-wide scans.\n",
        "\n",
        "8: *minGQ30*- Filtered genotypes to retain only those with a Genotype Quality (GQ) ≥ 30, removing uncertain genotype calls. This helps prevent the misclassification of homozygous vs heterozygous states.\n",
        "\n",
        "9: *hwe_0.05*- Removed variants deviating significantly from Hardy-Weinberg Equilibrium (p < 0.05), which could indicate genotyping errors or population substructure.\n",
        "\n",
        "10: *imiss_0.6*- Filtered out individuals with >60% missing genotypes to maintain data quality. Highly incomplete samples can distort PCA and admixture results.\n",
        "\n",
        "11: *miss_0.6*- Removed variants missing in >60% of individuals. This ensures each retained SNP is well represented across individuals.\n",
        "\n",
        "12: *mid95percentile* - Likely indicates the removal of extreme outliers in genotype depth or quality, retaining the middle 95% of the data. This removes the Lowest 2.5% of undercovered, unreliable sites and the highest 2.5% of multi-mapped, repetitive, or duplicated regions.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "IsZLf94VPkZ2"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Prerequisites and setup:\n",
        "To run this variant-filtering command line workflow, make sure you have Miniconda already installed, and it is working. Must have already downloaded the required input VCF file from Zenodo into your working directory.\n",
        "The setup involves creating and activating the required conda environments for vcftools, ggplot2, and base R, installing these packages via the bioconda and Cond forge channels."
      ],
      "metadata": {
        "id": "WQRRIryrPxk-"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bf840da6"
      },
      "source": [
        "First, we'll download and install Miniconda. We use `wget` to download the installer script and then execute it using `bash`."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Download Miniconda installer\n",
        "wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "\n",
        "# Install Miniconda silently into /usr/local/miniconda\n",
        "bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "# Add conda to PATH and initialize\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Accept Terms of Service\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Create environment with vcftools\n",
        "conda create -y -n vcftools -c bioconda vcftools"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ef02fc29"
      },
      "source": [
        "Next, we will list the packages installed in the `vcf_filter` environment to ensure `vcftools` is there."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "947ce0a6",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "conda list -n vcftools"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "c3d176aa",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Create the directory if it doesn't exist\n",
        "mkdir -p vcf_file\n",
        "\n",
        "# Download the VCF file into the created directory\n",
        "wget -P vcf_file https://zenodo.org/records/15173226/files/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "765a8796"
      },
      "source": [
        "First, we'll apply base quality (`--minQ`), genotype quality (`--minGQ`), and Hardy-Weinberg equilibrium (`--hwe`) filters to the VCF file. The output file name will reflect these new filters."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8c999e78",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Define input and output files\n",
        "input_vcf_gz=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3.recode.vcf.gz\"\n",
        "output_prefix_step1=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Run vcftools filters\n",
        "vcftools --gzvcf \"$input_vcf_gz\" \\\n",
        "         --minQ 30 \\\n",
        "         --minGQ 30 \\\n",
        "         --hwe 0.05 \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step1\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "22c1c9f3"
      },
      "source": [
        "Next, we will remove indels (insertions and deletions) from the filtered VCF file. The `--remove-indels` flag ensures that only SNP (Single Nucleotide Polymorphism) sites are retained. The output file name will again be updated to reflect this filter."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9e0f3796"
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Define input and output files\n",
        "input_vcf_step1=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_noIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "output_prefix_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Remove INDELs\n",
        "vcftools --vcf \"$input_vcf_step1\" \\\n",
        "         --remove-indels \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step2\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "40fc342c"
      },
      "source": [
        "Finally, we will apply the individual missingness filter. This command will generate an output file with a `.imiss` extension, which contains information about the fraction of missing sites for each individual."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "19b279ed",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Define input and output\n",
        "input_vcf_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "output_prefix_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05\"\n",
        "\n",
        "# Generate .imiss file\n",
        "vcftools --vcf \"$input_vcf_step2\" \\\n",
        "         --missing-indv \\\n",
        "         --out \"$output_prefix_step2\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d57d63f1"
      },
      "source": [
        "The command above will produce a file named `machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss`. You can inspect this file to see the fraction of missing sites for each individual in the `F_MISS` column. This information is crucial for identifying individuals with a high proportion of missing data, which might be excluded from further analysis if the missingness exceeds a certain threshold."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b19c2de6"
      },
      "source": [
        "Now, we will remove individuals that have a missing data proportion greater than 60%. This involves using `awk` to parse the `.imiss` file, identify individuals with `F_MISS` (fraction of missing data, which is the 5th column) greater than 0.6, and then passing these individual IDs to `vcftools` using the `--remove` flag."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "26579eef",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# File paths\n",
        "input_vcf_step2=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.recode.vcf\"\n",
        "imiss_file=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss\"\n",
        "output_prefix_step3=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6\"\n",
        "temp_remove_list=\"vcf_file/individuals_to_remove.txt\"\n",
        "\n",
        "# Extract individuals with >60% missing data in the temp file\n",
        "awk 'NR>1 && $5>0.6 {print $1}' \"$imiss_file\" > \"$temp_remove_list\"\n",
        "\n",
        "# Remove those individuals from the VCF\n",
        "vcftools --vcf \"$input_vcf_step2\" \\\n",
        "         --remove \"$temp_remove_list\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix_step3\"\n",
        "\n",
        "# Cleanup temporary file\n",
        "rm \"$temp_remove_list\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "956f9310"
      },
      "source": [
        "After this step, a new VCF file will be generated (prefixed with `machali_Aligned...imiss_0.6`) that excludes individuals with more than 60% missing data. This helps to improve the quality of downstream analyses by ensuring that only individuals with a sufficient amount of genotyped data are included."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now, We will apply SNP-level missingness filters (10–90%) to identify and remove low-quality variants and to determine the optimal threshold that keeps reliable SNPs while minimizing missing data.The (`--max-missing`) filter removes sites where more than X fraction of genotypes are missing."
      ],
      "metadata": {
        "id": "BTiCgWFwJRPV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Input VCF (after removing individuals >60% missing)\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6.recode.vcf\"\n",
        "\n",
        "# Loop over missingness thresholds (10% to 90%)\n",
        "for miss in {10..90..10}; do\n",
        "\n",
        "    # Convert percent to decimal\n",
        "    perc=$(echo \"scale=2; $miss / 100\" | bc)\n",
        "\n",
        "    # Output prefix for this missingness threshold\n",
        "    output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_${miss}\"\n",
        "\n",
        "    # Run VCFtools missingness filter\n",
        "    vcftools --vcf \"$input_vcf\" \\\n",
        "             --max-missing \"$perc\" \\\n",
        "             --recode \\\n",
        "             --out \"$output_prefix\"\n",
        "    echo \"Completed missingness filter: $miss% (max-missing $perc)\"\n",
        "done"
      ],
      "metadata": {
        "id": "UwkvZv2ress2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "This process generates nine new VCF files, each one filtered at a different level of allowed missingness. The output files are named according to the threshold, such as (`_miss_10.recode.vcf`), (`_miss_20.recode.vcf`), and so on. These files allow us to compare how strict or relaxed missingness filtering affects SNP retention, helping us decide the best cutoff for producing a clean and reliable dataset."
      ],
      "metadata": {
        "id": "AQhVawU4fcnk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "After this we have to count how many SNPs remain in each VCF file after applying different SNP-level missingness filters. For every thresholds, it extracts the number of SNPs and write the results into a text file (`variant_count.txt`)"
      ],
      "metadata": {
        "id": "qXCOzeXMNCxs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Remove old counts file if it exists\n",
        "rm -f variant_counts.txt\n",
        "\n",
        "# Loop over missingness thresholds\n",
        "for miss in {10..90..10}; do\n",
        "\n",
        "    # Input VCF for this threshold\n",
        "    vcf_file=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_${miss}.recode.vcf\"\n",
        "\n",
        "    # Count variants (non-header lines)\n",
        "    count=$(grep -vc \"^#\" \"$vcf_file\")\n",
        "\n",
        "    # Save: missingness  variant_count\n",
        "    echo \"${miss} ${count}\" >> vcf_file/variant_counts.txt\n",
        "\n",
        "    echo \"Counted variants for ${miss}% missingness → ${count}\"\n",
        "\n",
        "done"
      ],
      "metadata": {
        "collapsed": true,
        "id": "-0L6UqAhfd0m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Then, we create and activate (`ggplot2`) environment so that we can run R scripts and generate plots"
      ],
      "metadata": {
        "id": "8d0ZwfaYhgRU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Add conda to PATH\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Create the ggplot2 environment\n",
        "conda create -y -n ggplot2 -c conda-forge r-ggplot2\n",
        "\n",
        "# Activate the environment\n",
        "conda activate ggplot2"
      ],
      "metadata": {
        "collapsed": true,
        "id": "km2E_ExmhgmC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next step creates a visual QC plot that shows SNP retention chnages with different missigness thresholds, which will help to identify optimal missingness threshold for downstream analyses. This creates a clean line graph showing X-axis: missingness thresholds and Y-axis: number of SNPs retained at each thresholds"
      ],
      "metadata": {
        "id": "Q6JTdlbHib6E"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda ggplot2 environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate ggplot2\n",
        "\n",
        "# Create an R script to run ggplot2\n",
        "cat << 'EOF' > plot_variants.R\n",
        "library(ggplot2)\n",
        "library(scales)\n",
        "\n",
        "data <- read.table(\"vcf_file/variant_counts.txt\", header=FALSE,\n",
        "                   col.names=c(\"Missingness\", \"Variants\"))\n",
        "\n",
        "p <- ggplot(data, aes(x=Missingness, y=Variants)) +\n",
        "     geom_line(color=\"blue\") +\n",
        "     geom_point(color=\"red\") +\n",
        "     ggtitle(\"Number of Passed Variants vs. Missingness Filter\") +\n",
        "     xlab(\"Max Missingness (%)\") +\n",
        "     ylab(\"Number of Variants\") +\n",
        "     scale_y_continuous(labels = comma) +\n",
        "     scale_x_continuous(limits = c(10, 90)) +\n",
        "     theme_minimal()\n",
        "\n",
        "ggsave(\"vcf_file/variant_plot.png\", plot = p)\n",
        "EOF\n",
        "\n",
        "# Run the R script\n",
        "Rscript plot_variants.R"
      ],
      "metadata": {
        "id": "Hwyundz6icRV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next step calculates the mean sequencing depth for every SNP to ensure that variants are supported by a sufficient number of reads. For this purpose, we use the VCFtools command (`--site-mean-depth`), which calculates the average read depth for every genomic position across all individuals in the dataset. Here we choose (`machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf`) file for calculating mean depth."
      ],
      "metadata": {
        "id": "E1P6tH8kSupe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Use the correct miss_60 file\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\"\n",
        "\n",
        "output_prefix=\"vcf_file/depth\"\n",
        "\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --site-mean-depth \\\n",
        "         --out \"$output_prefix\"\n"
      ],
      "metadata": {
        "collapsed": true,
        "id": "9UdDgXgwjYBz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next steps prepares an R environment and uses it to calculate the middle 95% of the site-mean depth distribution for all SNPs in the dataset. First, we create and activate a dedicated Conda environment (R_env) that contains a clean installation of R. Next, we generate a small R script (compute_depth_range.R) that reads the file depth.ldepth.mean, which was previously produced using the VCFtools --site-mean-depth command. The R script extracts the MEAN_DEPTH column and calculates the 2.5th and 97.5th percentiles of the depth distribution using the quantile() function."
      ],
      "metadata": {
        "id": "dspFE9IClDej"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Add conda to PATH and initialize shell\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Create R environment (run once)\n",
        "conda create -y -n R_env -c conda-forge r-base\n",
        "\n",
        "# Activate R environment\n",
        "conda activate R_env\n",
        "\n",
        "# Create an R script to calculate middle 95% depth\n",
        "cat << 'EOF' > compute_depth_range.R\n",
        "depth_data <- read.table(\"vcf_file/depth.ldepth.mean\", header = TRUE)\n",
        "\n",
        "depths <- depth_data$MEAN_DEPTH\n",
        "\n",
        "lower_bound <- quantile(depths, 0.025, na.rm = TRUE)\n",
        "upper_bound <- quantile(depths, 0.975, na.rm = TRUE)\n",
        "\n",
        "cat(\"Middle 95% range:\", lower_bound, \"to\", upper_bound, \"\\n\")\n",
        "EOF\n",
        "\n",
        "# Run the R script\n",
        "Rscript compute_depth_range.R\n"
      ],
      "metadata": {
        "collapsed": true,
        "id": "UME3l06plD2d"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "By determining this middle 95% range, we establish data-driven thresholds (min-depth and max-depth) that will be used in the next filtering step to remove SNPs with suspicious or abnormal sequencing depth. This improves data quality and ensures that downstream population genomic analyses are based on reliable variants."
      ],
      "metadata": {
        "id": "52HTOEzClp30"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This step uses the depth percentiles calculated earlier to filter SNPs based on their mean sequencing depth, ensuring that only well-supported variants are retained in the dataset. After activating the vcf_filter environment, the script takes the VCF file that has already been cleaned for base quality, genotype quality, HWE, INDEL removal, and individual-level missingness filtering (specifically the version filtered at 60% missingness). Using the middle 95% depth range—defined by the 2.5th percentile (12.83) and 97.5th percentile (24.66) of site-mean depth—it applies VCFtools’ --min-meanDP and --max-meanDP commands to remove SNPs that fall outside this normal depth range."
      ],
      "metadata": {
        "id": "_b2QF-fPnmpH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda vcf_filter environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Input VCF from 60% missingness threshold\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60.recode.vcf\"\n",
        "\n",
        "# Middle 95% Depth Percentiles (calculated earlier)\n",
        "min_depth=12.8302\n",
        "max_depth=24.6604\n",
        "\n",
        "# Output prefix\n",
        "output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile\"\n",
        "\n",
        "# Run VCFtools to filter by mean depth\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --min-meanDP \"$min_depth\" \\\n",
        "         --max-meanDP \"$max_depth\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix\""
      ],
      "metadata": {
        "id": "ix2BDe2tnmy-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next command searches the .imiss file for individuals whose sample IDs contain the substring “ZSB”, extracts their IDs, and saves them into a separate text file. By using grep 'ZSB', we locate all rows corresponding to samples whose names include “ZSB,” which typically represent a specific group of individuals"
      ],
      "metadata": {
        "id": "3ehU7lukpden"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "grep 'ZSB' vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05.imiss | awk '{print $1}' > vcf_file/zsb_samples.txt"
      ],
      "metadata": {
        "id": "fMAxlS62pd4t"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "YmnXhyNo5PGO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda vcf_filter environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate vcftools\n",
        "\n",
        "# Define input & output\n",
        "input_vcf=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile.recode.vcf\"\n",
        "remove_list=\"vcf_file/zsb_samples.txt\"\n",
        "\n",
        "output_prefix=\"vcf_file/machali_Aligned_rangeWideMerge_strelka_update2_BENGAL_mac3_passOnly_biallelicOnly_rmvIndels_minMAF0Pt05_chr_E2_minDP3_minQ30_minGQ30_hwe_0.05_imiss_0.6_miss_60_mid95percentile_noZSB\"\n",
        "\n",
        "# Run vcftools to remove ZSB individuals\n",
        "vcftools --vcf \"$input_vcf\" \\\n",
        "         --remove \"$remove_list\" \\\n",
        "         --recode \\\n",
        "         --out \"$output_prefix\"\n",
        "\n",
        "echo \"Done: ZSB samples removed. Output = ${output_prefix}.recode.vcf\"\n"
      ],
      "metadata": {
        "id": "x0aQbZIa5hFC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Tutorial session"
      ],
      "metadata": {
        "id": "TJ1HdfnQZJUI"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "lt7dJEuQZNlx"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}