{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Pranav-Datar/pacbio_denovo2/blob/main/NGS_collab_Linux_basics.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tptE67QQFzih"
      },
      "source": [
        "# **This hands-on workshop is designed to introduce students' to the basic analysis pipeline for Next-Generation Sequencing (NGS) reads, specifically from the Illumina platform for population genetics analysis.**\n",
        "We will focus on whole genome resequencing data (WGS), though the same pipeline may also be applied to RAD-seq or amplicon-seq data (with minor modifications) if a reference genome is available.\n",
        "\n",
        "By the end of the workshop, participants will understand how to:\n",
        "\n",
        "\n",
        "\n",
        "*   Work with raw sequencing data (FASTQ files)\n",
        "*   Perform quality control and preprocessing\n",
        "\n",
        "\n",
        "*   Map reads to a reference genome\n",
        "*   Call and filter genetic variants (SNPs/indels)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "abW1jFslGzHs"
      },
      "source": [
        "# Pre-requisites & Setup\n",
        "Before joining the workshop, please make sure you are familiar with Unix/Linux commands and tools which is what we will be primarily using .\n",
        "Go through the command description listed as below, to get a good overview of the command language we will be using in this workshop."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DH3Ji3OtXXS_"
      },
      "source": [
        "In this Linux cheat sheet, we will cover all the most important Linux commands, from the basics to the advanced.\n",
        "1. `ls` : List files and directories.\\\n",
        "Examples:\\\n",
        "`ls -l`\n",
        "Displays files and directories with detailed information.\\\n",
        "`ls -a`\n",
        "Shows all files and directories, including.\\\n",
        "`ls -lh`\n",
        "Displays file sizes in a human-readable format.\n",
        "\n",
        "2. `cd` : Change directory.\\\n",
        "`cd /path/to/directory`\n",
        "Changes the current directory to the specified path.\n",
        "\n",
        "3. `pwd` : Print current working directory.\n",
        "\n",
        "4. `mkdir` : Create a new directory.\n",
        "\n",
        "5. `rm` : Remove directory\\\n",
        "`rm -r` Delete the directory and it's contents\n",
        "\n",
        "6. `mv` : Move or rename the file.\n",
        "\n",
        "7. `cp`: Copy files and directories.\n",
        "\n",
        "8. `cat` : To view the content of the file.\n",
        "\n",
        "9. `head` : To view the first few lines of the file.\n",
        "\n",
        "10. `tail` : To view the last few lines of the file.\n",
        "\n",
        "11. `ln` : Creates links between files.\n",
        "\n",
        "12. `find` : Search for files and directory.\n",
        "\n",
        "13. `chmod`: Change file permissions.\n",
        "\n",
        "14. `tar`: Create or extract archive files.\n",
        "\n",
        "15. `gzip`: Compress files\n",
        "\n",
        "16. `zip`: Create compressed `zip` archives.\n",
        "\n",
        "17. `ssh`: Securely connect to remote server.\\\n",
        "`ssh user@hostname`\n",
        "Initiates an SSH connection to the specified hostname.\n",
        "\n",
        "18. `scp` : Securely copy files between hosts.\\\n",
        "`scp file.txt user@hostname:/path/to/destination`\n",
        "Securely copies “file.txt” to the specified remote host.\n",
        "\n",
        "19. `wget` : Download files from the web\\\n",
        "`wget http://example.com/file.txt`\n",
        "Downloads “file.txt” from the specified URL."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "D9WYLbMxObfq"
      },
      "source": [
        "# To get the most out of this workshop, we recommend checking out the following resource:\n",
        "How Illumina Sequencing works?\\\n",
        "[Check out this brief and engaging video to understand the sequencing technology behind your data](https://youtu.be/fCd6B5HRaZ8?si=GPKriSYOUKEPGFoo)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tGd84_zjHskR"
      },
      "source": [
        "# Question 1\n",
        "\n",
        "Create the following table in a file named test_table.txt on terminal\n",
        "\n",
        "```\n",
        "1  2  3  4  5\n",
        "6  7  8  9  10\n",
        "11  12  13  14\n",
        "15  16  17  18\n",
        "```\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MDynvBNNvuzp",
        "outputId": "793690c6-b0d6-414a-f2ff-ade8eb20c9f2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " 1   2   3   4   5\n",
            " 6   7   8   9  10\n",
            "11  12  13  14\n",
            "15  16  17  18\n"
          ]
        }
      ],
      "source": [
        "#Question 1 solution\n",
        "\n",
        "%%bash\n",
        "touch test_table.txt #creates a new text file in the directory\n",
        "{\n",
        "    printf \"%2d  %2d  %2d  %2d  %2d\\n\" 1 2 3 4 5\n",
        "    printf \"%2d  %2d  %2d  %2d  %2d\\n\" 6 7 8 9 10\n",
        "    printf \"%2d  %2d  %2d  %2d\\n\" 11 12 13 14\n",
        "    printf \"%2d  %2d  %2d  %2d\\n\" 15 16 17 18\n",
        "} > test_table.txt # \">\" forwards the entered table data into test_table.txt\n",
        "\n",
        "cat test_table.txt # This will display the content of the file"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SkPpeZtfIvon"
      },
      "source": [
        "#Question 2\n",
        "\n",
        "Create a directory named genomics_workshop_PoODL and move the file test_table.txt into it\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "7O1QGc_4w8f6"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "mkdir -p genomics_workshop #creates new directory or folder named \"genomics_workshop\" only if it doesn't exist\n",
        "\n",
        "# Check if test_table.txt exists in the current directory before attempting to move it\n",
        "if [ -f \"test_table.txt\" ]; then\n",
        "    mv test_table.txt genomics_workshop #moves the test_table.txt file into the genomics_workshop folder\n",
        "fi"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QwVpyWkcJp_a"
      },
      "source": [
        "#Question 3\n",
        "\n",
        "Count the number of lines in test_table.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "htTqLPbF0PdF",
        "outputId": "ec8c2526-3724-4e57-e578-661c2d139607"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "4 genomics_workshop/test_table.txt\n"
          ]
        }
      ],
      "source": [
        "#Question 3 solution\n",
        "\n",
        "%%bash\n",
        "wc -l genomics_workshop/test_table.txt # (wc - wordcount; l- lines) - specify the full path to the file"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "n23I40fCJuX-"
      },
      "source": [
        "#Question 4\n",
        "\n",
        "Count the number of columns per row in test_table.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "C1NWV89W1wRK",
        "outputId": "5b3f0c73-32a5-407b-d258-e4c60482d057"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "5\n",
            "5\n",
            "4\n",
            "4\n"
          ]
        }
      ],
      "source": [
        "#Question 4 solution\n",
        "\n",
        "%%bash\n",
        "awk '{print NF}' genomics_workshop/test_table.txt #awk reads input line by line. {print NF}: Executes for each line of the input file."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NQ-Ajy64JzT8"
      },
      "source": [
        "#Question 5\n",
        "\n",
        "Count the number of times \"1\" appears in test_table.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "B7j5BtyH2jV7",
        "outputId": "77cca8a0-5582-4ee4-9f5b-1fff52bb90fa"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "11\n"
          ]
        }
      ],
      "source": [
        "#Question 5 solution\n",
        "\n",
        "%%bash\n",
        "grep -o '1' genomics_workshop/test_table.txt | wc -l"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EP-FM4ehJ3MI"
      },
      "source": [
        "#Question 6\n",
        "\n",
        "Count the number of words/strings containing \"1\" in test_table.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iWlXgvxx2uwK",
        "outputId": "a3192c5c-bc20-44c1-e9fd-aa8688b7685e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "10\n"
          ]
        }
      ],
      "source": [
        "#Question 6 solution\n",
        "\n",
        "%%bash\n",
        "grep -o '\\S*1\\S*' genomics_workshop/test_table.txt | wc -l\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Tutorial Questions**\n",
        "\n",
        "1) From an ecological genomics perspective, why is learning the command line (and not just GUIs) important?\n",
        "2) What are the main limitations of amplicon-seq for population genomics compared with WGS?\n",
        "3) What does a pipe command \"|\" do?\n",
        "4) What is “sequencing-by-synthesis” (SBS)?\n",
        "5) What role do fluorescent tags on nucleotides play in Illumina sequencing?\n"
      ],
      "metadata": {
        "id": "2hDV0IfBiKo_"
      }
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}